### PR TITLE
Update slate, refactor plugins and add more typing

### PR DESCRIPTION
**Pull request by Atyanar**
First of all: **Sorry** for the giant PR. IF you read on you will understand why so much changed...

This PR had one main objective: Updating slate to the newest version. Since the changelog of slate is incomplete and the code broke on several occasions without error messages, I also decided to convert all plugins to typescript to get an idea of what changed. This means that the diff is super huge, and I am sorry for that. However, it is also a change I've been meaning to implement sooner or later anyway.

## Major Changes

The most important change to slate is that the change object is gone. Instead, every function now gets the editor object directly, which is API equivalent in most occasions. However, this also means that changes from outside the editor are not really feasible anymore. Therefore, the plusmenu and textmenu moved into the editor as plugins. They will still be rendered beside the editor through the renderEditor hook, though. This also separates the concerns more nicely, so no more moving refs around. Instead plugins render themselves when they are included (e.g. plusmenu), no further changes necessary. 

A second major breaking change is that every hook now also gets a next function. If you don't return this new function at the end of your function handling, no other hook will get called and you will be left with very weird behaviour.

## Minor Changes
LastSaved is now saved in a React.Context.

This PR only refactors. It does not add any new functionality. I tested every functionality by hand, so nothing should be seriously broken. 

The typings are not 100% perfect everywhere. However, since there were no typings before it does not make things worse, and the typings can easily be specified later on, if necessary.

In some occasions, I also converted .css files to Styled Components. However, this is also not complete (which is fine with the same reasoning as with typescript).

I converted most Components to PureComponents to prevent unnecessary rerenders, since the Editor as parent rerenders on every change.

### CODE REVIEW

**Comment by FWirtz** [comment]
cool thanks! Esp. `change` being gone sounds exciting. 👍

File: docs/guides/create-plugin.mdx [hunk 1]
```diff
@@ -53,74 +55,73 @@ const RenderImageNode = {...}
...
-}
+    return next();
+  },
+};
 ```
 
 Our `RenderImageNode` is called whenever a node should be rendered. It checks if the type of the node to be rendered is an `img` and if so renders our custom `<ImageNode ... />`.
 
-If that is not the case nothing is returned, thus allowing other plugins that override `renderNode()` to render the current node.
+If that is not the case next() is returned, thus allowing other plugins that override `renderNode()` to render the current node.
```
Comment (FWirtz) [inline docs/guides/create-plugin.mdx]:
```suggestion
If that is not the case `next()` is returned, thus allowing other plugins that override `renderNode()` to render the current node.
```

File: docs/structure/plugin.mdx [hunk 2]
```diff
@@ -51,13 +52,14 @@ const RenderImageNode = {
 ```
 
 ## Requiring plugins
-All plugins are located in the `plugins/` folder and manually required in `Editor.jsx`.
+
+All plugins are located in the `plugins/` folder and exported there as a list in the index.ts.
```
Comment (FWirtz) [inline docs/structure/plugin.mdx]:
```suggestion
All plugins are located in the `plugins/` folder and exported there as a list in the `index.ts`.
```

File: docs/guides/create-plugin.mdx [hunk 3]
```diff
@@ -53,74 +55,73 @@ const RenderImageNode = {...}
...
 That's it! All it takes to create a custom plugin that renders images is to
 
 1. Create a custom image component
-2. Define a change to insert that image
-3. Override renderNode
+2. Override renderNode
+3. Insert a block with the image type somewhere and see slate work it's magic
 
-> **Note**: Our custom change is called inside `plugins/plus-menu/actions.js:onClickImageButton` which is an EventListener on the "Add image" button.
+> **Note**: Our custom insertImage function is called inside `plugins/plus-menu/actions.js:onClickImageButton` which is an EventListener on the "Add image" button.
```
Comment (FWirtz) [inline docs/guides/create-plugin.mdx]:
```suggestion
> **Note**: Our custom `insertImage` function is called inside `plugins/plus-menu/actions.js:onClickImageButton` which is an EventListener on the "Add image" button.
```

**Comment by Atyanar** [comment]
damn @FWirtz , adding your markdown changes removed your approval ^^

Also, I cannot merge this. Could you do it or give me the appropriate rights? Or does @janrenz do this?

**Comment by janrenz** [comment]
Let me know if I should merge. Lgtm

**Comment by FWirtz** [comment]
ah yeah makes sense @Atyanar. You should have admin rights now?! 😊

**Comment by FWirtz** [comment]
@janrenz can not wait :D

**Pull request merged**